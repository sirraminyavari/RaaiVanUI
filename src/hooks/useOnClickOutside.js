import { useEffect } from 'react';

/**
 * This hook allows you to detect clicks outside of a specified element.
 * @hook
 * @param {Object} ref
 * @param {function} handler
 */
const useOnClickOutside = (ref, handler) => {
  useEffect(
    () => {
      const listener = (e) => {
        //! Do nothing if clicking ref's element or descendent elements
        if (!ref.current || ref.current.contains(e.target)) {
          return;
        }
        handler(e);
      };

      document.addEventListener('mousedown', listener);
      document.addEventListener('touchstart', listener);

      return () => {
        document.removeEventListener('mousedown', listener);
        document.removeEventListener('touchstart', listener);
      };
    },
    // Add ref and handler to effect dependencies
    // It's worth noting that because passed in handler is a new ...
    // ... function on every render that will cause this effect ...
    // ... callback/cleanup to run every render. It's not a big deal ...
    // ... but to optimize you can wrap handler in useCallback before ...
    // ... passing it into this hook.
    [ref, handler]
  );
};

export default useOnClickOutside;
