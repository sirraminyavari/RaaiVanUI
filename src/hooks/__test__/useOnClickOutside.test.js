import { useRef } from 'react';
import { render, cleanup, screen, fireEvent } from '@testing-library/react';

import useOnClickOutside from '../useOnClickOutside';

describe('useOnClickOtside', () => {
  let wrapper;
  const hookCallback = jest.fn();

  afterEach(cleanup);

  const renderHook = (hook) => {
    const HookWrapper = () => {
      const ref = useRef();
      hook(ref, hookCallback);

      return (
        <div ref={ref} data-testid="ref">
          Container to click outside
        </div>
      );
    };

    wrapper = render(<HookWrapper />);
  };

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('works', () => {
    renderHook(useOnClickOutside);

    const ref = screen.getByTestId('ref');

    fireEvent.mouseDown(ref);

    expect(hookCallback).not.toBeCalled();

    fireEvent.mouseDown(document);

    expect(hookCallback).toBeCalled();

    wrapper.debug();
  });
});
