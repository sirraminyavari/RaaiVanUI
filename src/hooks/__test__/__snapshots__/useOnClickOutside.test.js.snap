// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useOnClickOtside matches snapshot 1`] = `undefined`;
